Separate config.2

meejah opened the pull request:
Updated version (post-1.13-release) of #467

meejah commented:
This is a re-based version of #467 (including all fixes from that review). There was only one conflict, so should be good-to-go (in case anyone has comments, I'll avoid merging for a bit .. also I'd like to merge *after* #507

exarkun commented on src/allmydata/node.py:
```diff
@@ -232,6 +276,92 @@ def get_config(self, section, option, default=_None, boolean=False):
+            value = fileutil.read(privname)
+        except EnvironmentError:
+            if os.path.exists(privname):
+                raise
```
Race condition here.  Checking for ENOENT in the exception is better.

exarkun commented on src/allmydata/node.py:
```diff
@@ -232,6 +276,92 @@ def get_config(self, section, option, default=_None, boolean=False):
+            return fileutil.read(privname).strip()
+        except EnvironmentError:
+            if os.path.exists(privname):
+                raise
```
Should be an ENOENT check here as well

exarkun commented on src/allmydata/test/test_client.py:
```diff
@@ -212,6 +213,19 @@ def test_reserved_bad(self):
                            "reserved_space = bogus\n")
         self.failUnlessRaises(ValueError, client.create_client, basedir)
 
+    def test_web_apiauthtoken(self):
```
A docstring here would be great.

exarkun commented on src/allmydata/test/test_client.py:
```diff
@@ -228,6 +242,21 @@ def test_web_staticdir(self):
 
     # TODO: also test config options for SFTP.
 
+    def test_ftp_create(self):
```
Docstring here too

exarkun commented on src/allmydata/test/test_introducer.py:
```diff
@@ -40,9 +41,14 @@ def test_backwards_compat_import(self):
         from allmydata.introducer import IntroducerNode
         IntroducerNode  # pyflakes
 
+    def test_create(self):
```
docstring here too

exarkun commented on src/allmydata/test/test_node.py:
```diff
@@ -120,9 +120,8 @@ def test_tahoe_cfg_utf8(self):
-        n.setServiceParent(self.parent)
-        self.failUnlessEqual(n.get_config("node", "nickname").decode('utf-8'),
+        config = read_config(basedir, "")
+        self.failUnlessEqual(config.get_config("node", "nickname").decode('utf-8'),
```
new code should use `assert...` style methods instead of the `fail...` style methods.  Up to you if you want to also make that change here or not (since this is not really a new line, just changed).

exarkun commented on src/allmydata/test/test_node.py:
```diff
@@ -133,8 +132,57 @@ def test_tahoe_cfg_hash_in_name(self):
+        config = read_config(basedir, "")
+        self.failUnless(config.nickname == nickname)
+
+    def test_config_required(self):
```
docstring please :)

exarkun commented on src/allmydata/test/test_node.py:
```diff
@@ -133,8 +132,57 @@ def test_tahoe_cfg_hash_in_name(self):
+        with self.assertRaises(Exception):
+            config.get_config_from_file("it_does_not_exist", required=True)
+
+    def test_private_config_unreadable(self):
```
docstring here as well

exarkun commented on src/allmydata/test/test_node.py:
```diff
@@ -133,8 +132,57 @@ def test_tahoe_cfg_hash_in_name(self):
+            # We don't know how to test that unprivileged users can't read this
+            # thing.  (Also we don't know exactly how to set the permissions so
+            # that unprivileged users can't read this thing.)
+            raise unittest.SkipTest("We don't know how to set permissions on Windows.")
```
The condition here is entirely static so it could be computed before the test is run and the skip attribute or decorator could be used instead.

exarkun commented on src/allmydata/test/test_node.py:
```diff
@@ -133,8 +132,57 @@ def test_tahoe_cfg_hash_in_name(self):
+            # We don't know how to test that unprivileged users can't read this
+            # thing.  (Also we don't know exactly how to set the permissions so
+            # that unprivileged users can't read this thing.)
+            raise unittest.SkipTest("We don't know how to set permissions on Windows.")
```
same comments here as for test_private_config_unreadable

exarkun commented on src/allmydata/test/test_node.py:
```diff
@@ -133,8 +132,57 @@ def test_tahoe_cfg_hash_in_name(self):
+        with self.assertRaises(Exception):
+            config.get_private_config("foo")
+
+    def test_private_config_missing(self):
```
docstring please

exarkun commented on src/allmydata/test/test_node.py:
```diff
@@ -207,123 +271,164 @@ def call_setLogDir(logdir):
+        self.basedir = self.mktemp()
+        create_node_dir(self.basedir, "testing")
+
+    def test_0(self):
```
A better name and a docstring would be great here.  Same comment for the following tests.

meejah commented on src/allmydata/client.py:
```diff
@@ -38,7 +36,7 @@
 
 def _valid_config_sections():
-    cfg = _common_config_sections()
+    cfg = node._common_config_sections()
```
It might just be me habitually putting "_" in front now so I don't have to support the API without thinking about it further ;)
(I've been using this to mean "definitely internal stuff" but yeah it's not documented for Tahoe-LAFS :/ )

meejah commented on src/allmydata/client.py:
```diff
@@ -574,15 +597,17 @@ def init_sftp_server(self):
 
             try:
-                magic_folders = magic_folder.load_magic_folders(self.basedir)
+                magic_folders = magic_folder.load_magic_folders(self.config._basedir)
```
https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2939#ticket

meejah commented:
Okay I believe I've addressed all the comments from the review (thanks @exarkun !). One thing I didn't change was "leading underscore" usage...but we should probably have a better idea about when to do that or not. I'm happy to remove the leading underscores if you think that's more-consistent..?

meejah commented:
(Dang, github sure makes you feel bad about "dismiss review" ... :/ )

meejah commented on src/allmydata/client.py:
```diff
@@ -154,14 +161,32 @@ def stopService(self):
         return service.Service.stopService(self)
 
 
+def read_config(basedir, portnumfile, generated_files=[]):
```
Hmm, I'm thinking maybe we should add a "confirm all new functions/methods have a docstring" to codechecks.....

meejah commented:
If there are more fixups, I am happy to do them as a followup. Merging this now to make some progress on "async-initializers" and other branches.
